plc4j/integrations/apache-nifi/nifi-plc4x-processors/src/main/java/org/apache/plc4x/nifi/record/Plc4xReadResponseRecordSet.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.nifi.serialization.record.RecordSchema;
+import org.apache.nifi.serialization.record.RecordSet;
+import org.apache.plc4x.java.api.messages.PlcReadResponse;
+import org.apache.plc4x.nifi.util.PLC4X_DATA_TYPE;
+import org.apache.plc4x.nifi.util.PLC4X_PROTOCOL;
+import org.apache.plc4x.nifi.util.Plc4xCommon;
+import org.slf4j.LoggerFactory;
+
+
+public class Plc4xReadResponseRecordSet implements RecordSet, Closeable {
+    private static final Logger logger = LoggerFactory.getLogger(Plc4xReadResponseRecordSet.class);
+    private final PlcReadResponse readResponse;
+    private final RecordSchema schema;
+    private final Set<String> rsColumnNames;
+    private boolean moreRows;
+
+
+    public Plc4xReadResponseRecordSet(final Map<String, String> plcAddressMap, final PlcReadResponse readResponse, final RecordSchema readerSchema, PLC4X_PROTOCOL PROTOCOL) throws IOException {
+        this.readResponse = readResponse;
+        moreRows = true;
+        this.schema = createSchema(plcAddressMap, readerSchema, true, PROTOCOL);
+        rsColumnNames = plcAddressMap.keySet();
+        
+    }
+
+    @Override
+    public RecordSchema getSchema() {
+        return schema;
+    }
+
+    // Protected methods for subclasses to access private member variables
+    protected PlcReadResponse getReadResponse() {
+        return readResponse;
+    }
+
+    protected boolean hasMoreRows() {
+        return moreRows;
+    }
+
+    protected void setMoreRows(boolean moreRows) {
+        this.moreRows = moreRows;
+    }
+
+    @Override
+    public Record next() throws IOException {
+        if (moreRows) {
+             final Record record = createRecord(readResponse);
+             setMoreRows(false);
+             return record;
+        } else {
+             return null;
+        }
+    }
+
+    @Override
+    public void close() {
+        //do nothing
+    }
+
+    protected Record createRecord(final PlcReadResponse readResponse) throws IOException{
+        final Map<String, Object> values = new HashMap<>(schema.getFieldCount());
+
+        for (final RecordField field : schema.getFields()) {
+            final String fieldName = field.getFieldName();
+
+            final Object value;
+            if (rsColumnNames.contains(fieldName)) {
+                value = normalizeValue(readResponse.getObject(fieldName));
+            } else {
+                value = null;
+            }
+
+            values.put(fieldName, value);
+        }
+
+        //TODO add timestamp field to schema
+        values.put(Plc4xCommon.PLC4X_RECORD_TIMESTAMP_FIELD_NAME, System.currentTimeMillis());
+        	
+        return new MapRecord(schema, values);
+    }
+
+    @SuppressWarnings("rawtypes")
+    private Object normalizeValue(final Object value) {
+        if (value == null) {
+            return null;
+        }
+        if (value instanceof List) {
+            return ((List) value).toArray();
+        }
+        return value;
+    }
+
```

Comment:
As stated on the list, I think creating the schema once, when the first data is retrieved *from* the data would be best